modular_chomp/code/modules/vote/vote_datum.dm

```diff
@@ -0,0 +1,200 @@
+
+        if(VOTE_RESULT_TYPE_SKEWED)
+            var/required_votes = ceil(length(voted) * 0.7)  // 70% of total votes
+            if(maxvotes >= required_votes && length(winning_options) == 1)
+                var/res = winning_options[1]
+                if(res in choices)
+                    to_chat(world, span_interface("<b><code>[res]</code> won the vote with a 70% majority!</b>"))
+                    return res
+                else
+                    to_chat(world, span_interface("The winner of the vote ([sanitize(res)]) isn't a valid choice? What the heck?"))
+                    stack_trace("Vote concluded with an invalid answer. Answer was [sanitize(res)], choices were [json_encode(choices)]")
+                    return null
+
+            to_chat(world, span_interface("<b>No option received 70% of the votes. The vote did not pass.</b>"))
+            return null
+
+    return null
+
+/datum/vote/proc/announce(start_text, var/time = vote_time)
+    to_chat(world, span_lightpurple("Type <b>vote</b> or click <a href='?src=\ref[src];[HrefToken()];vote=open'>here</a> to place your vote. \
+        You have [time] seconds to vote."))
+    world << sound('sound/ambience/alarm4.ogg', repeat = 0, wait = 0, volume = 50, channel = 3)
+
+/datum/vote/Topic(href, list/href_list)
+    if(href_list["vote"] == "open")
+        if(src)
+            tgui_interact(usr)
+        else
+            to_chat(usr, "There is no active vote to participate in.")
+
+/datum/vote/proc/tick()
+	if(remaining() == 0)
+		var/result = calculate_result()
+		handle_result(result)
+		qdel(src)
+
+/datum/vote/Destroy(force)
+	if(SSvote.active_vote == src)
+		SSvote.active_vote = null
+	return ..()
+
+/datum/vote/proc/handle_result(result)
+	return
+
+/datum/vote/proc/generate_choices()
+	return
+
+/*
+	UI STUFFS
+*/
+
+/datum/vote/tgui_state(mob/user)
+	return GLOB.tgui_always_state
+
+/datum/vote/tgui_interact(mob/user, datum/tgui/ui = null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "VotePanel", "Vote Panel")
+		ui.open()
+
+/datum/vote/tgui_data(mob/user)
+	var/list/data = list()
+	data["remaining"] = remaining()
+	data["user_vote"] = null
+	if(user.ckey in voted)
+		data["user_vote"] = voted[user.ckey]
+
+	data["question"] = question
+	data["choices"] = choices
+
+	if(show_counts || check_rights(R_ADMIN, FALSE, user))
+		data["show_counts"] = TRUE
+
+		var/list/counts = list()
+		for(var/ck in voted)
+			if(voted[ck] in counts)
+				counts[voted[ck]]++
+			else
+				counts[voted[ck]] = 1
+
+		data["counts"] = counts
+	else
+		data["show_counts"] = FALSE
+		data["counts"] = list()
+
+	return data
+
+/datum/vote/tgui_act(action, list/params, datum/tgui/ui, datum/tgui_state/state)
+	if(..())
+		return
+
+	. = TRUE
+
+	switch(action)
+		if("vote")
+			if(params["target"] in choices)
+				voted[usr.ckey] = params["target"]
+			else
+				message_admins(SPAN_WARNING("User [key_name_admin(usr)] spoofed a vote in the vote panel!"))
```

Comment:
best to use the same spans everywhere span_warning non capital, I will nuke the double defines sometime